checkmate/models/db/mixins.py

```diff
@@ -0,0 +1,117 @@
+        )
+
+    @classmethod
+    def find_matches(cls, session, hex_hashes, limit=None):
+        """Find matching rules for the specified hashes.
+
+        :param session: DB session to execute within
+        :param hex_hashes: List of URL hashes to find
+        :param limit: Limit the number of returned values
+        :return: Iterable of matching CustomRule objects
+        """
+
+        query = session.query(cls).filter(cls.hash.in_(hex_hashes))
+
+        if limit:
+            query = query.limit(limit)
+
+        return query
+
+
+class BulkUpsertMixin:
+    """A mixin for models that want to support bulk upserting."""
+
+    BULK_UPSERT_INDEX_ELEMENTS = None
+    """Elements to detect collisions on."""
+
+    BULK_UPSERT_UPDATE_ELEMENTS = None
+    """Elements to set in the event of a collision."""
+
+    BLOCK_SIZE = 10000
+    """Numer of elements to attempt to update in one go."""
+
+    @classmethod
+    def bulk_upsert(cls, session, values):
+        """Create or update a number of rows at once.
+
+        This will match on the "rule" portion and must include "hash" and
+        "tags". This will not delete any rows.
+
+        :param session: DB session to execute within
+        :param values: A list of dicts of columns to update
+        :raise NotImplementedError: If you have not provided the expected class
+            values
+        """
+
+        if not cls.BULK_UPSERT_INDEX_ELEMENTS or not cls.BULK_UPSERT_UPDATE_ELEMENTS:
+            raise NotImplementedError(
+                "You must provide the correct elements for this mixin to work"
+            )
+
+        total_items = 0
+        for block in cls._chunk(values):
+            total_items += len(block)
+
+            cls._bulk_upsert(
+                session,
+                block,
+                index_elements=cls.BULK_UPSERT_INDEX_ELEMENTS,
+                update_elements=cls.BULK_UPSERT_UPDATE_ELEMENTS,
+            )
+
+    @classmethod
+    def _chunk(cls, values):
+        block = []
+
+        for value in values:
+            block.append(value)
+
+            if len(block) >= cls.BLOCK_SIZE:
+                yield block
+                block = []
+
+        if block:
+            yield block
+
+    @classmethod
+    def _bulk_upsert(cls, session, values, index_elements, update_elements):
+        stmt = insert(cls).values(values)
+
+        stmt = stmt.on_conflict_do_update(
+            # Match when the rules are the same
+            index_elements=index_elements,
+            # Then set these elements
+            set_={
+                element: getattr(stmt.excluded, element) for element in update_elements
+            },
+        )
+
+        session.execute(stmt)
+
+        # Let SQLAlchemy know that something has changed, otherwise it will
+        # never commit the transaction we are working on and it will get rolled
+        # back
+        mark_changed(session)
```

Comment:
In the spirit of keeping models simple and keeping code and logic out of the models, and moving it into services instead, this looks to `BulkUpsertMixin` might want to change it into a service in the future, instead of a model class mixin. Maybe something like: `bulk_upsert_svc.upsert(CustomRule, values)`

Reply:
Yeah maybe. There's currently no services layer, but we could add one. It would get around some of the awkward passing of the DB session, but would leave a question as to how to hook them up correctly with the knowledge of which columns to do.

It does seem very plain SQL stuff so far, so I think it's reasonable to have it in the DB layer. The other one is a bit more usage oriented, but still basically runs a very simple query for you.